File: archctl/src/doctor.ts
type Severity = "ok" | "warn" | "fail";

interface Finding {
  id: string;
  title: string;
  severity: Severity;
  detail: string;
}

export interface DoctorReport {
  ok: boolean;
  findings: Finding[];
  profilOpencode: string;
  xdg: { data: string; config: string };
  renderers: { structurizr: boolean; plantuml: boolean; kroki: boolean };
}

function probe(url: string, timeoutSec = 2): boolean {
  const r = spawnSync(
    "curl",
    ["-sS", "-o", "/dev/null", "-w", "%{http_code}", "--max-time", String(timeoutSec), url],
    { encoding: "utf8" },
  );
  if (r.status !== 0) return false;
  const code = Number((r.stdout ?? "0").trim());
  return code >= 200 && code < 400;
}

function probeBinary(name: string): boolean {
  const r = spawnSync(name, ["--version"], { encoding: "utf8" });
  return r.status === 0;
}

import { spawnSync } from "node:child_process";
import { existsSync } from "node:fs";
import { join } from "node:path";

function xdgPaths(): { data: string; config: string } {
  const home = process.env.HOME ?? "/tmp";
  const data = process.env.XDG_DATA_HOME ?? `${home}/.local/share`;
  const config = process.env.XDG_CONFIG_HOME ?? `${home}/.config`;
  return { data: join(data, "archctl"), config: join(config, "archctl") };
}

export function runDoctor(): DoctorReport {
  const findings: Finding[] = [];
  const xdg = xdgPaths();
  const xdgOk = existsSync(xdg.data) || (() => {
    try {
      const r = spawnSync("mkdir", ["-p", xdg.data]);
      return r.status === 0;
    } catch {
      return false;
    }
  })();
  findings.push({
    id: "xdg.data",
    title: "XDG data directory",
    severity: xdgOk ? "ok" : "warn",
    detail: xdg.data,
  });
  findings.push({
    id: "xdg.config",
    title: "XDG config directory",
    severity: existsSync(xdg.config) ? "ok" : "warn",
    detail: xdg.config,
  });

  const structurizr = probe("http://localhost:18080/") || probeBinary("structurizr");
  findings.push({
    id: "renderer.structurizr",
    title: "Structurizr (local)",
    severity: structurizr ? "ok" : "warn",
    detail: structurizr
      ? "reachable (HTTP localhost:18080 or structurizr CLI on PATH)"
      : "not reachable; run `podman run -d --rm --name archctl-structurizr -p 18080:8080 structurizr/structurizr:latest`",
  });

  const plantuml = probeBinary("plantuml");
  const kroki = probe("http://localhost:18000/");
  const plantumlOk = plantuml || kroki;
  findings.push({
    id: "renderer.plantuml",
    title: "PlantUML (local)",
    severity: plantumlOk ? "ok" : "warn",
    detail: plantuml
      ? "plantuml CLI on PATH"
      : kroki
      ? "local Kroki on localhost:18000"
      : "not reachable; install `plantuml` or `podman run -d --rm --name archctl-kroki -p 18000:8000 yuzutech/kroki:latest`",
  });

  const opencode = probeBinary("opencode");
  const archctl = probeBinary("archctl");
  findings.push({
    id: "opencode.cli",
    title: "OpenCode CLI",
    severity: opencode ? "ok" : "warn",
    detail: opencode ? "opencode on PATH" : "opencode not on PATH",
  });
  findings.push({
    id: "archctl.cli",
    title: "archctl CLI",
    severity: archctl ? "ok" : "warn",
    detail: archctl ? "archctl on PATH" : "archctl not on PATH (run from `archctl/` via npm run)",
  });

  const ok = findings.every((f) => f.severity !== "fail");
  return {
    ok,
    findings,
    profilOpencode: process.env.OPENCODE_CONFIG_DIR ?? "<unset>",
    xdg,
    renderers: { structurizr, plantuml: plantumlOk, kroki },
  };
}

const args = process.argv.slice(2);
const human = args.includes("--human");
const r = runDoctor();
if (human) {
  console.log("archctl doctor");
  for (const f of r.findings) console.log(`  [${f.severity.toUpperCase().padEnd(4)}] ${f.id}: ${f.title} — ${f.detail}`);
  console.log(r.ok ? "DOCTOR: OK" : "DOCTOR: FAIL");
} else {
  console.log(JSON.stringify(r, null, 2));
}
process.exit(r.ok ? 0 : 1);

File: archctl/src/doctor.rs
use crate::environment::Environment;
use crate::identity::{identity_summary, resolve_source_identity};
use crate::scope::{check_all_scopes, render_report_line, ScopeCheckReport};
use crate::xdg::{resolve_xdg, user_home};
use std::process::Command;
use tracing::{info, warn};

#[derive(Debug)]
pub struct Finding {
    pub id: String,
    pub detail: String,
    pub severity: Severity,
}

#[derive(Debug, PartialEq, Eq)]
pub enum Severity {
    Ok,
    Warn,
    Fail,
}

pub fn run() -> Result<i32, anyhow::Error> {
    let layout = resolve_xdg();
    let mut findings: Vec<Finding> = Vec::new();
    // The Environment port is the boundary. doctor.rs is an internal
    // entry point — it uses SystemEnvironment directly. Tests can
    // inject a context via a future `run_with_env` if needed.
    let cwd = crate::environment::SystemEnvironment.current_dir().ok();
    let cwd_str = cwd
        .as_ref()
        .map(|p| p.to_string_lossy().to_string())
        .unwrap_or_default();

    for (id, path) in [
        ("xdg.data", &layout.data),
        ("xdg.config", &layout.config),
        ("xdg.state", &layout.state),
        ("xdg.cache", &layout.cache),
    ] {
        let ok = path.exists() || std::fs::create_dir_all(path).is_ok();
        findings.push(Finding {
            id: id.to_string(),
            detail: path.display().to_string(),
            severity: if ok { Severity::Ok } else { Severity::Fail },
        });
    }

    findings.push(http_finding("renderer.structurizr", "http://localhost:18080/"));
    findings.push(http_finding("renderer.plantuml", "http://localhost:18000/"));
    findings.push(binary_finding("opencode.cli", "opencode"));
    findings.push(binary_finding("archctl.cli", "archctl"));

    println!("archctl doctor");
    let identity = resolve_source_identity(&cwd_str);
    info!(home = %user_home().display(), "doctor starting");
    for f in &findings {
        let tag = match f.severity {
            Severity::Ok => "OK  ",
            Severity::Warn => "WARN",
            Severity::Fail => "FAIL",
        };
        println!("  [{tag}] {}: {}", f.id, f.detail);
    }
    println!("  sourceIdentity: {}", identity_summary(&identity));
    let failed = findings.iter().filter(|f| f.severity == Severity::Fail).count();
    if failed > 0 {
        warn!(failures = failed, "doctor detected failures");
        println!("DOCTOR: FAIL");
        Ok(1)
    } else {
        info!("doctor: all OK");
        println!("DOCTOR: OK");
        Ok(0)
    }
}

fn http_finding(id: &str, url: &str) -> Finding {
    let probe = Command::new("curl")
        .args(["-sS", "-o", "/dev/null", "-w", "%{http_code}", "--max-time", "2", url])
        .output();
    let ok = match probe {
        Ok(o) => o.status.success() && String::from_utf8_lossy(&o.stdout).trim().starts_with('2'),
        Err(_) => false,
    };
    Finding {
        id: id.to_string(),
        detail: if ok { format!("reachable ({url})") } else { format!("not reachable ({url})") },
        severity: if ok { Severity::Ok } else { Severity::Warn },
    }
}

fn binary_finding(id: &str, name: &str) -> Finding {
    let probe = Command::new(name).arg("--version").output();
    let ok = matches!(&probe, Ok(o) if o.status.success());
    let detail = match probe {
        Ok(o) if o.status.success() => String::from_utf8_lossy(&o.stdout)
            .lines()
            .next()
            .unwrap_or("")
            .to_string(),
        _ => "not on PATH".to_string(),
    };
    Finding {
        id: id.to_string(),
        detail,
        severity: if ok { Severity::Ok } else { Severity::Warn },
    }
}

/// Run the scope gates against every manifest under `manifests/`.
/// Returns the exit code (0 if all gates pass or no manifests
/// exist; 1 if any gate fails).
/// Designed to be called from `archctl doctor --check-scope` but is
/// also useful from tests.
pub fn check_scope(cwd: &std::path::Path) -> Result<i32, anyhow::Error> {
    let manifests_dir = cwd.join("manifests");
    if !manifests_dir.exists() {
        println!("(no manifests/ directory at {})", cwd.display());
        println!("SCOPE: OK (no scopes declared)");
        return Ok(0);
    }
    let reports = check_all_scopes(cwd)?;
    print_scope_reports(&reports);
    let failed = reports.iter().filter(|r| !r.passed()).count();
    if failed > 0 {
        println!("SCOPE: FAIL");
        Ok(1)
    } else if reports.is_empty() {
        println!("(no manifests under manifests/)");
        println!("SCOPE: OK (no scopes declared)");
        Ok(0)
    } else {
        println!("SCOPE: OK");
        Ok(0)
    }
}

/// Render a list of scope check reports in a way that is human-
/// readable. The summary line is per-scope; full findings go to
/// stderr for tools that want detail.
fn print_scope_reports(reports: &[ScopeCheckReport]) {
    for report in reports {
        println!("{}", render_report_line(report));
        for f in &report.findings {
            eprintln!(
                "    [{}] {}: {}",
                match f.severity {
                    crate::scope::ScopeSeverity::Fail => "FAIL",
                    crate::scope::ScopeSeverity::Warn => "WARN",
                },
                f.gate.name(),
                f.message
            );
        }
    }
}
